packages/dev/core/src/Misc/deepMerger.ts

```diff
@@ -0,0 +1,26 @@
+// https://stackoverflow.com/a/48218209
+/**
+ * Merges a series of objects into a single object, deeply.
+ * @param objects The objects to merge (objects later in the list take precedence).
+ * @returns The merged object.
+ */
+export function deepMerge<T extends Object>(...objects: T[]): T {
+    const isRecord = (obj: unknown): obj is Record<string, unknown> => !!obj && typeof obj === "object";
+
+    return objects.reduce<Record<string, unknown>>((prev, obj) => {
+        Object.keys(obj).forEach((key) => {

```

Comment:
We have deep copy already implemented, would that be something you can reuse here? Also, reduce and forEach are very readable, but not as performant as a for-in or recursive calls. Is that a problem here?

This kind of implementation might be faster:

```javascript
const isObject = (item) => {
    return item && typeof item === 'object' && !Array.isArray(item);
};

const mergeDeep = (target, ...sources) => {
    if (!sources.length) return target;

    const source = sources.shift();

    if (isObject(target) && isObject(source)) {
        for (const key in source) {
            if (isObject(source[key])) {
                if (!target[key]) Object.assign(target, { [key]: {} });
                mergeDeep(target[key], source[key]);
            } else {
                Object.assign(target, { [key]: source[key] });
            }
        }
    }

    return mergeDeep(target, ...sources);
};
```

But maybe we should run perf tests before

Reply:
Side note - i noticed the usage for this here is for relatively small objects, so perf would not be an issue, however, if it will ever be used on larger objects, we should be sure that it works as fast as we can.

Reply:
I saw deep copy, but I didn't really understand it's purpose as it doesn't seem to actually copy deeply. It copies the top level properties, and if the property values are objects, it clones them, but it doesn't merge them. If I'm missing something here, let me know!

Regarding perf, I think it is ok for the current use, we could always improve it in the future. Mostly I just saw we already had the code so I just moved it without making any runtime changes (just some improvements to typing).